### PR TITLE
feat: Transitivity Detection for Supply Chain

### PR DESCRIPTION
**Pull request by mmcqd**
This PR adds support for detection of whether a dependency is direct or transitive to Semgrep Supply Chain. Currently only support python and JS. It works by reading the manifest file associated with a given lockfile.
PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: cli/src/semdep/parse_lockfile.py [hunk 1]
```diff
@@ -21,7 +23,7 @@
 
 logger = getLogger(__name__)
 
-from semgrep.semgrep_interfaces.semgrep_output_v0 import FoundDependency
+from semgrep.semgrep_interfaces.semgrep_output_v0 import FoundDependency, Transitivity
```
Comment (brendongo) [inline cli/src/semdep/parse_lockfile.py]: You import Transitivity twice

File: cli/src/semdep/parse_lockfile.py [hunk 2]
```diff
@@ -45,38 +51,40 @@ def extract_npm_lockfile_hash(s: str) -> Dict[str, List[str]]:
...
-            parsed_name = "@".join(first_quoted.split("@")[:-1])
-            return parsed_name
+def parse_yarn(
+    lockfile_text: str, manifest_text: Optional[str]
+) -> Generator[FoundDependency, None, None]:
+    def version_sources(line: str) -> Tuple[str, List[Tuple[str, str]]]:
+        constraint_strs = line[:-1].split(",")
+        constraints = []
+        for c in constraint_strs:
+            c = c.strip().strip('"')
```
Comment (brendongo) [inline cli/src/semdep/parse_lockfile.py]: Does .strip("\" ") work or is it possible there are spaces after "

File: cli/src/semdep/parse_lockfile.py [hunk 3]
```diff
@@ -45,38 +51,40 @@ def extract_npm_lockfile_hash(s: str) -> Dict[str, List[str]]:
...
-        if '"' not in line:
-            return line.split("@")[0]
-        else:
-            first_quoted = "".join(line.split('"')[1:2])
-            parsed_name = "@".join(first_quoted.split("@")[:-1])
-            return parsed_name
+def parse_yarn(
+    lockfile_text: str, manifest_text: Optional[str]
+) -> Generator[FoundDependency, None, None]:
+    def version_sources(line: str) -> Tuple[str, List[Tuple[str, str]]]:
```
Comment (brendongo) [inline cli/src/semdep/parse_lockfile.py:57]: Would be helpful to have a comment of what you expect line to look like (and also a quick comment on what this does cause it's not immediately clear from function name

File: cli/src/semdep/parse_lockfile.py [hunk 4]
```diff
@@ -45,38 +51,40 @@ def extract_npm_lockfile_hash(s: str) -> Dict[str, List[str]]:
...
+        constraints = []
+        for c in constraint_strs:
+            c = c.strip().strip('"')
+            if c[0] == "@":
+                name, constraint = c[1:].split("@")
+                name = "@" + name
+            else:
+                name, constraint = c.split("@")
+            constraints.append((name, constraint))
+        return constraints[0][0], constraints
```
Comment (brendongo) [inline cli/src/semdep/parse_lockfile.py:72]: Would returning Version/SemVer type here work or does it need to be strings?

File: cli/src/semdep/parse_lockfile.py [hunk 5]
```diff
@@ -89,30 +97,43 @@ def remove_quotes(s: str) -> str:
...
     # Newer versions of NPM (>= v7) use 'packages'
     # But 'dependencies' is kept up to date, and 'packages' uses relative, not absolute names
     # https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json
     if "dependencies" in as_json:
         deps = as_json["dependencies"]
-    elif "packages" in as_json:
-        deps = as_json["packages"]
     else:
-        logger.info("Found package-lock with no 'dependencies' or 'packages'")
+        logger.info("Found package-lock with no 'dependencies'")
```
Comment (brendongo) [inline cli/src/semdep/parse_lockfile.py]: Should this be info or debug?

File: cli/src/semdep/parse_lockfile.py [hunk 6]
```diff
@@ -344,45 +407,64 @@ def parse_dep(line: str) -> Optional[FoundDependency]:
     yield from (dep for dep in deps if dep)
 
 
-def parse_Poetry_str(lockfile_text: str) -> Generator[FoundDependency, None, None]:
+def parse_poetry(
+    lockfile_text: str, manifest_text: Optional[str]
+) -> Generator[FoundDependency, None, None]:
+    # poetry.lock file are not quite valid TOML >:(
```
Comment (brendongo) [inline cli/src/semdep/parse_lockfile.py:417]: :(

File: cli/src/semdep/parse_lockfile.py (hunk 1 above)
Comment (brendongo) [reply cli/src/semdep/parse_lockfile.py]: I also cant find these Classes in the atd repo

File: cli/src/semgrep/target_manager.py [hunk 7]
```diff
@@ -725,7 +726,16 @@ def get_lockfile_dependencies(
         lockfiles = self.get_files_for_language(ecosystem).kept
         parsed: Dict[Path, List[FoundDependency]] = {}
         for lockfile in lockfiles:
-            deps = parse_lockfile_str(lockfile.read_text(encoding="utf8"), lockfile)
+            path, lockfile_pattern = lockfile.parents[0], lockfile.parts[-1]
```
Comment (brendongo) [inline cli/src/semgrep/target_manager.py]: nit: indexing always makes me nervous but it's probably fine

**Review by brendongo — APPROVED** [review]

File: cli/src/semdep/parse_lockfile.py (hunk 5 above)
Comment (mmcqd) [reply cli/src/semdep/parse_lockfile.py]: probably debug

File: cli/src/semdep/parse_lockfile.py (hunk 4 above)
Comment (mmcqd) [reply cli/src/semdep/parse_lockfile.py:72]: I don't think making them Versions would do anything. If "^1.0.0" appears in the manifest then "^1.0.0" will appear in the lockfile. So string equality vs semver equality will be the same.

File: cli/src/semdep/parse_lockfile.py (hunk 1 above)
Comment (mmcqd) [reply cli/src/semdep/parse_lockfile.py]: I need to make a PR there

File: cli/src/semdep/parse_lockfile.py (hunk 2 above)
Comment (mmcqd) [reply cli/src/semdep/parse_lockfile.py]: no that should work

**Pull request merged**